commit 9e33a9f37e19e13564831f998285766fcbc33c64
Author: amollberg <amollberg@users.noreply.github.com>
Date:   Mon Aug 1 20:22:43 2016 +0200

    Renamed list-hunks.py -> list_hunks.py

diff --git a/list-hunks.py b/list_hunks.py
similarity index 100%
rename from list-hunks.py
rename to list_hunks.py

commit 667611ce644c9975c5de026402ec0563ee31a09b
Author: amollberg <amollberg@users.noreply.github.com>
Date:   Mon Aug 1 21:26:14 2016 +0200

    Use list_hunks.py in console_ui

diff --git a/console_ui.py b/console_ui.py
index 5c7a4cf..5f1ac32 100644
--- a/console_ui.py
+++ b/console_ui.py
@@ -0,0 +1,3 @@
+#!/usr/bin/env python
+# encoding: utf-8
+#import os
@@ -2,0 +6,8 @@ from generate_matrix import *
+from list_hunks import get_diff
+
+NPYSCREEN_AVAILABLE = False
+try:
+  from curses_ui import *
+  NPYSCREEN_AVAILABLE = True
+except ImportError:
+  print "Curses unavailable; using plain text."
@@ -24,0 +36,9 @@ def print_hunkogram(diff_list):
+def display_hunkogram_screen(diff_list):
+  matrix = generate_matrix(diff_list)
+  hash_width = 8
+  App = HunkogramApp()
+  App._diff_list = diff_list
+  App._matrix = matrix
+  App.run()
+
+
@@ -27 +47 @@ def main():
-  lines = [line.rstrip() for line in sys.stdin]
+  lines = get_diff('HEAD~4..HEAD')
@@ -30,2 +50,6 @@ def main():
-  print_hunkogram(diff_list)
-
+  print diff_list
+  return
+  if NPYSCREEN_AVAILABLE:
+    display_hunkogram_screen(diff_list)
+  else:
+    print_hunkogram(diff_list)

commit 67aa22e3a76e752102aab31a8b411bea27b8419c
Author: amollberg <amollberg@users.noreply.github.com>
Date:   Mon Aug 1 21:28:27 2016 +0200

    Fix: Allow patch without fragments

diff --git a/parse_patch.py b/parse_patch.py
index 91cf5fe..88f0503 100644
--- a/parse_patch.py
+++ b/parse_patch.py
@@ -145 +145 @@ class FilePatchHeader():
-    while lines[0] != '' and lines[0][0:4] != '--- ':
+    while lines and lines[0] != '' and lines[0][0:4] != '--- ':
@@ -148 +148 @@ class FilePatchHeader():
-    if lines[0][0:4] != '--- ':
+    if not lines or lines[0][0:4] != '--- ':

commit 5906fe349c1d19d9a3534e04b467e0de80a8f071
Author: amollberg <amollberg@users.noreply.github.com>
Date:   Mon Aug 1 21:29:27 2016 +0200

    Added failing test for empty/rename patch bug

diff --git a/test/main.py b/test/main.py
index cd77c25..3ae0b7b 100644
--- a/test/main.py
+++ b/test/main.py
@@ -67,0 +68,7 @@ class Test(unittest.TestCase):
+  def test_002_004(self):
+    self.check_diffs(['003-add-one-line-to-empty-file.diff',
+                      '002-rename-empty-file.diff',
+                      '004-remove-one-line-empty-file.diff'],
+                     ['#.',
+                      '..',
+                      '#.'])
